tests/samplers_tests/test_grid.py

```diff
@@ -0,0 +1,165 @@
+from collections import OrderedDict
+import itertools
+import numpy as np
+import pytest
+
+import optuna
+from optuna import samplers
+
+if optuna.type_checking.TYPE_CHECKING:
+    from collections import ValuesView  # NOQA
+    from typing import Dict  # NOQA
+    from typing import List  # NOQA
+    from typing import Union  # NOQA
+
+    from optuna.samplers.grid import GridValueType  # NOQA
+    from optuna.trial import Trial  # NOQA
+
+
+def _n_grids(search_space):
+    # type: (Dict[str, List[Union[str, float, None]]]) -> int
+
+    return int(np.prod([len(v) for v in search_space.values()]))
+
+
+def test_study_optimize_with_single_search_space():
+    # type: () -> None
+
+    def objective(trial):
+        # type: (Trial) -> float
+
+        a = trial.suggest_int('a', 0, 100)
+        b = trial.suggest_uniform('b', -0.1, 0.1)
+        c = trial.suggest_categorical('c', ('x', 'y'))
+        d = trial.suggest_discrete_uniform('d', -5, 5, 1)
+        e = trial.suggest_loguniform('e', 0.0001, 1)
+
+        if c == 'x':
+            return a * d
+        else:
+            return b * e
+
+    # Test that all combinations of the grid is sampled.
+    search_space = {
+        'b': np.arange(-0.1, 0.1, 0.05),
+        'c': ['x', 'y'],
+        'd': [-0.5, 0.5],
+        'e': [0.1],
+        'a': list(range(0, 100, 20)),
+    }
+    n_grids = _n_grids(search_space)
+    study = optuna.create_study(sampler=samplers.GridSampler(search_space))
+    study.optimize(objective, n_trials=n_grids)
+
+    def sorted_values(d):
+        # type: (Dict[str, List[GridValueType]]) -> ValuesView[List[GridValueType]]
+
+        return OrderedDict(sorted(d.items())).values()
+
+    all_grids = itertools.product(*sorted_values(search_space))
+    all_suggested_values = [tuple([p for p in sorted_values(t.params)]) for t in study.trials]
+    assert set(all_grids) == set(all_suggested_values)
+
+    ids = sorted([t.system_attrs['grid_id'] for t in study.trials])
+    assert ids == list(range(n_grids))
+
+    # Test that an optimization fails if the number of trials is more than that of all grids.
+    with pytest.raises(ValueError):
+        study.optimize(objective, n_trials=1)
```

Comment:
[my question] Is this comment correct?

I think the code contradicts the comment as I guess `n_grids` is greater than 1.

Reply:
I understood. The study already runs `n_grids` trials.